File: 20-actors/watari/methods/test_ingest.py
#!/usr/bin/env python3
"""Tests for the watari 渡り offline AIS/ADS-B normalizer (methods/ingest.py).

    PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python3 -m pytest test_ingest.py
    python3 test_ingest.py

Covers offline normalization (public-broadcast → :craft/:craft.fix datoms, all
:representative) AND the G7 outward gate: a live network fetch is REFUSED unless the
operator attestation env var is set.
"""
from __future__ import annotations

import json
import os
import pathlib
import sys

try:
    import ingest
    from ingest import normalize
except ImportError:
    from watari.methods import ingest  # type: ignore
    from watari.methods.ingest import normalize  # type: ignore

_BATCH = pathlib.Path(__file__).resolve().parent.parent / "data" / "ingest" / "sample-batch.json"


def _batch():
    return json.loads(_BATCH.read_text())


def test_normalize_emits_vessel_and_aircraft_craft():
    craft, fixes = normalize(_batch())
    kinds = {c[":craft/kind"] for c in craft.values()}
    assert ":vessel" in kinds and ":aircraft" in kinds
    assert len(fixes) >= len(craft)            # at least one fix per craft


def test_normalized_records_are_representative():
    craft, fixes = normalize(_batch())
    assert all(c.get(":craft/sourcing") == ":representative" for c in craft.values())
    assert all(f.get(":craft.fix/sourcing") == ":representative" for f in fixes)


def test_fix_carries_source_tag():
    craft, fixes = normalize(_batch())
    sources = {f.get(":craft.fix/source") for f in fixes}
    assert sources <= {":ais", ":adsb"}        # only public-broadcast sources


def test_g4_no_person_fields_in_normalized_output():
    craft, fixes = normalize(_batch())
    for rec in list(craft.values()) + fixes:
        for k in rec:
            assert ":person" not in k          # craft, never a person (G4)


def test_g7_live_fetch_refused_without_operator_gate():
    # default mode (no WATARI_OPERATOR_GATE) must REFUSE --live
    saved = os.environ.pop("WATARI_OPERATOR_GATE", None)
    try:
        raised = False
        try:
            ingest.main(["ingest.py", "--live"])
        except SystemExit as e:
            raised = True
            assert "G7" in str(e) or "refus" in str(e).lower() or "gate" in str(e).lower()
        assert raised, "--live must refuse without the operator gate"
    finally:
        if saved is not None:
            os.environ["WATARI_OPERATOR_GATE"] = saved


def _run():
    fns = [v for k, v in sorted(globals().items())
           if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"watari ingest.py: {len(fns)}/{len(fns)} tests passed")
    return True


if __name__ == "__main__":
    sys.exit(0 if _run() else 1)

File: 20-actors/watari/methods/test_ingest.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_ingest.py.
(ns watari.methods.test-ingest
  "Tests for the watari 渡り offline AIS/ADS-B normalizer (methods/ingest.clj).

  Covers offline normalization (public-broadcast → :craft/:craft.fix records, all
  :representative) AND the G7 outward gate (live network fetch refused without operator gate).

  Run:  bb --classpath 20-actors 20-actors/watari/methods/test_ingest.clj"
  (:require [watari.methods.ingest :as ing]
            [cheshire.core :as json]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [clojure.test :refer [deftest is run-tests]]))

(def ^:private this-file *file*)
(defn- batch []
  (json/parse-string
   (slurp (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile
              (io/file "data" "ingest" "sample-batch.json")))
   true))

(deftest normalize-emits-vessel-and-aircraft-craft
  (let [{:keys [craft fixes]} (ing/normalize (batch))
        kinds (set (map :craft/kind (vals craft)))]
    (is (contains? kinds :vessel))
    (is (contains? kinds :aircraft))
    (is (>= (count fixes) (count craft)))))      ; at least one fix per craft

(deftest normalized-records-are-representative
  (let [{:keys [craft fixes]} (ing/normalize (batch))]
    (is (every? #(= (:craft/sourcing %) :representative) (vals craft)))
    (is (every? #(= (:craft.fix/sourcing %) :representative) fixes))))

(deftest fix-carries-source-tag
  (let [{:keys [fixes]} (ing/normalize (batch))
        sources (set (map :craft.fix/source fixes))]
    (is (clojure.set/subset? sources #{:ais :adsb}))))   ; only public-broadcast sources

(deftest g4-no-person-fields-in-normalized-output
  (let [{:keys [craft fixes]} (ing/normalize (batch))]
    (doseq [rec (concat (vals craft) fixes)
            k (keys rec)]
      (is (not (str/includes? (str k) ":person")) "craft, never a person (G4)"))))

(deftest counts-match-sample
  (let [{:keys [craft fixes]} (ing/normalize (batch))]
    (is (= (count craft) 4))     ; 2 vessels + 2 aircraft
    (is (= (count fixes) 4))))

(deftest g7-live-fetch-refused-without-operator-gate
  (is (some? (ing/live-refusal ["--live"] nil)) "--live refused without the operator gate")
  (is (str/includes? (ing/live-refusal ["--live"] nil) "G7"))
  (is (str/includes? (ing/live-refusal ["--live"] "1") "not implemented")) ; gated → R0 scaffold
  (is (nil? (ing/live-refusal ["--batch" "x.json"] nil))))                 ; offline not refused

(when (= *file* (System/getProperty "babashka.file"))
  (require 'clojure.set)
  (let [{:keys [fail error]} (run-tests 'watari.methods.test-ingest)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
